Code/GraphMol/SmilesParse/smatest.cpp

```diff
@@ -2671,11 +2671,81 @@ void testGithub2142() {
 
   BOOST_LOG(rdInfoLog) << "done" << std::endl;
 }
+
+void testGithub2565() {
+  BOOST_LOG(rdInfoLog) << "-------------------------------------" << std::endl;
+  BOOST_LOG(rdInfoLog)
+      << "Testing Github #2565: Chirality reversed on SMARTS generation."
+      << std::endl;
+
+  std::vector<std::string> smiles({
+      R"(O=C1C[C@]([H])1F)",
+      R"(Cl[C@@H]1CCC1=O)",
+      R"(N1CCN=C1[C@H]1CCCc2ccccc21)",
+      R"(C[C@@H](Cl)[C@H]1CC[C@@H](Cl)CC1)",
+      R"(Fc1cn([C@@H]2CCCO2)c(=O)[nH]c1=O)",
+      R"([C@@]1(C)(C(C)(C)C)O[C@@H](CN)[C@H](C[NH3+])O1)",
+
+      // these are Ok
+      R"(O=C1C[C@](Cl)1F)",
+      R"(Br[C@@H](Cl)F)"});
+
+  for (const auto &smi : smiles) {
+    auto *mol = SmilesToMol(smi);
+    const std::string smarts = MolToSmarts(*mol, true);
+    auto *query = SmartsToMol(smarts);
+
+    std::vector<MatchVectType> matches;
+    TEST_ASSERT(SubstructMatch(*mol, *query, matches, true, true, true, false));
+  }
+
+  BOOST_LOG(rdInfoLog) << "done" << std::endl;
+}
+
+void testGithub2568() {
+  BOOST_LOG(rdInfoLog) << "-------------------------------------" << std::endl;
+  BOOST_LOG(rdInfoLog) << "Testing Github #2568: SMARTS parser reverses "
+                          "chirality on leading chiral atoms. "
+                       << std::endl;
+
+  std::vector<std::string> smiles({
+      R"(O=C1C[C@]([H])1F)",
+      R"([C@H](O)(F)Cl)",
+
+      /*
+       * This is a tough one: it won't keep chirality in SMILES.
+       * Let's skip it for now.
+       * R"([C@](O)(F)Cl)",
+       */
```

Comment:
Just an info point: this loses chirality in the SMILES because the sanitization "knows" that a three-coordinate neutral C can't be chiral. So it gets removed.
Molecules built from SMARTS aren't sanitized on parsing, so this doesn't happen.

Reply:
Yeah, I imagined that. But my though was that, since we have 3 specified non-H neighbors and a chirality set, it would be reasonable to assume that the missing neighbor is an implicit H, and add it to the mol.

Reply:
I meant in SMILES, SMARTS should understand it as a fragment, and not add the implicit H.

Reply:
No, because by specifying the carbon with `[C@]`  you have explicitly said that it does *not* have any implicit Hs. That's how you specify radicals in SMILES.

so `[C@](O)(F)Cl` in SMILES corresponds to a C that has a radical electron.

Reply:
Oh, I see. I totally forgot about radicals!